src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -0,0 +1,526 @@
+async def coursier_resolve_lockfile(
+    coursier: Coursier,
+    maven_requirements: MavenRequirements,
+) -> CoursierResolvedLockfile:
+    """Run `coursier fetch ...` against a list of Maven coordinates and capture the result.
+
+    This rule does two things in a single Process invocation:
+
+        * Runs `coursier fetch` to let Coursier do the heavy lifting of resolving
+          dependencies and downloading resolved artifacts (jars, etc).
+        * Copies the resolved artifacts into the Process output directory, capturing
+          the artifacts as content-addressed `Digest`s.
+
+    It's important that this happens in the same process, since the process isn't
+    guaranteed to run on the same machine as the rule, nor is a subsequent process
+    invocation.  This guarantees that whatever Coursier resolved, it was fully
+    captured into Pants' content addressed artifact storage.
+
+    Note however that we still get the benefit of Coursier's "global" cache if it
+    had already been run on the machine where the `coursier fetch` runs, so rerunning
+    `coursier fetch` tends to be fast in practice.
+
+    Finally, this rule bundles up the result into a `CoursierResolvedLockfile`.  This
+    data structure encapsulates everything necessary to either materialize the
+    resolved dependencies to a classpath for Java invocations, or to write the
+    lockfile out to the workspace to hermetically freeze the result of the resolve.
+    """
+
+    if len(maven_requirements) == 0:
+        return CoursierResolvedLockfile(entries=())
+
+    coursier_report_file_name = "coursier_report.json"
+    process_result = await Get(
+        ProcessResult,
+        Process(
+            argv=[
+                "/bin/sh",
+                coursier.wrapper_script,
+                coursier.coursier.exe,
+                coursier_report_file_name,
+                *list(maven_requirements),
+            ],
+            input_digest=coursier.digest,
+            output_directories=("classpath",),
+            output_files=(coursier_report_file_name,),
+            description="Run coursier resolve",
+            level=LogLevel.DEBUG,
+        ),
+    )
+    report_digest = await Get(
+        Digest, DigestSubset(process_result.output_digest, PathGlobs([coursier_report_file_name]))
+    )
+    report_contents = await Get(DigestContents, Digest, report_digest)
+    report = json.loads(report_contents[0].content)
+
+    lockfile_entries = []
+    for dep in report["dependencies"]:
+        file_path = PurePath(dep["file"])
+        classpath_dest = f"classpath/{file_path.name}"
+        artifact_digest = await Get(
+            Digest, DigestSubset(process_result.output_digest, PathGlobs([classpath_dest]))
+        )
+        stripped_artifact_digest = await Get(Digest, RemovePrefix(artifact_digest, "classpath"))
+        artifact_file_digest = await Get(
+            FileDigest, ExtractFileDigest(stripped_artifact_digest, file_path.name)
+        )
```

Comment:
This can be parallelized via MultiGet. Note that you are doing many Gets within a for loop, which is almost never the right thing to do. Move those `awaits` out of the loop.

You will possibly need to use `zip()`.

Reply:
I couldn't see a way to do this without adding an entirely new helper rule to facilitate the parallelization, so that's what I did.  Let me know if I missed a simpler approach.

Reply:
It will look something like:

```python
artifact_digests = await MultiGet(Get(Digest, DigestSubset(process_result.output_digest, PathGlobs([classpath_dest]) for classpath_dest in report["dependencies"])
stripped_artifact_digests = await MultiGet(Get(Digest, RemovePrefix(digest, "classpath")) for digest in artifact_digests)
```

And so on. You can then zip the final `artifcat_file_digests` collection with the `report["dependencies"]` collection to create the final `CoursierLockfileEntry` objects.

Does that make sense?

Reply:
Ah I see, that does make sense.

Reply:
Done